src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java

```diff
@@ -42,105 +42,116 @@ public int getLength() {
-                    throw new RuntimeException(String.format("Cannot apply soft clipping operator to the middle of a read: %s to be clipped at %d-%d", read.getName(), start, myStop));
-                }
+            default:
+                throw new IllegalStateException("Unexpected Clipping operator type " + algorithm);
+        }
+    }
+
+    private GATKRead applySOFTCLIP_BASES(final GATKRead readCopied) {
+        if (readCopied.isUnmapped()) {
+            // we can't process unmapped reads
+            throw new UserException("Read Clipper cannot soft clip unmapped reads");
+        }
 
-                final Cigar oldCigar = read.getCigar();
+        int myStop = stop;
+        if ((stop + 1 - start) == readCopied.getLength()) {
+            // BAM representation issue -- we can't SOFTCLIP away all bases in a read, just leave it alone
+            myStop--; // just decrement stop
+        }
 
-                int scLeft = 0, scRight = read.getLength();
-                if (start == 0) {
-                    scLeft = myStop + 1;
-                } else {
-                    scRight = start;
-                }
+        if (start > 0 && myStop != readCopied.getLength() - 1) {
+            throw new RuntimeException(String.format("Cannot apply soft clipping operator to the middle of a read: %s to be clipped at %d-%d", readCopied.getName(), start, myStop));
+        }
 
-                final Cigar newCigar = softClip(oldCigar, scLeft, scRight);
-                read.setCigar(newCigar);
+        final Cigar oldCigar = readCopied.getCigar();
 
-                final int newClippedStart = getNewAlignmentStartOffset(newCigar, oldCigar);
-                final int newStart = read.getStart() + newClippedStart;
-                read.setPosition(read.getContig(), newStart);
+        int scLeft = 0, scRight = readCopied.getLength();
+        if (start == 0) {
+            scLeft = myStop + 1;
+        } else {
+            scRight = start;
+        }
 
-                break;
+        final Cigar newCigar = softClip(oldCigar, scLeft, scRight);
+        readCopied.setCigar(newCigar);
 
-            case REVERT_SOFTCLIPPED_BASES:
-                read = revertSoftClippedBases(read);
-                break;
+        final int newClippedStart = getNewAlignmentStartOffset(newCigar, oldCigar);
+        final int newStart = readCopied.getStart() + newClippedStart;
+        readCopied.setPosition(readCopied.getContig(), newStart);
+        return readCopied;
+    }
 
-            default:
-                throw new IllegalStateException("Unexpected Clipping operator type " + algorithm);
+    private GATKRead applyWRITE_NS_Q0S(final GATKRead readCopied) {
+        final byte[] quals = readCopied.getBaseQualities();
+        final byte[] bases = readCopied.getBases();
+        final byte[] newBases = new byte[bases.length];
+        final byte[] newQuals = new byte[quals.length];
+        for (int i = 0; i < bases.length; i++) {
```

Comment:
changed to ` for (int i = start, n = Math.min(newBases.length, stop+1); i < n; i++) {`